contributor-docs/adrs/adr-020-live-regions.md

```diff
@@ -0,0 +1,234 @@
+
+## Decision
+
+In order to have a common interop point for live region announcements, Primer React will
+make use of a `live-region` custom element from `@primer/live-region-element`.
+This package will be included and published from the `primer/react` repo.
+
+The custom element exposes a way to make announcements that can be used
+across frameworks. This makes it useful not only for Primer but GitHub as a
+whole. The `@primer/live-region-element` exports two helpers to use for making
+announcements: `announce()` and `announceFromElement`. Both helpers can be used
+when working in Primer and by teams at GitHub.
+
+In addition, `@primer/react` will leverage and export the following helpers for
+use within Primer React and GitHub:
+
+- The `AriaStatus` component to correspond with `role="status"`
+- The `AriaAlert` component to correspond with `role="alert"`
+
+Within `@primer/react`, we should lint against usage of `aria-live` and the
+corresponding roles (if possible) and suggest using these alternatives instead.
+
+> [!NOTE]
+> Both `AriaStatus` and `AriaAlert` will trigger an announcement when the component is
+> rendered. As a result, they should only be used for dynamically rendered
+> content. Otherwise, they will trigger announcements on page load. In cases
+> where they should always be present, then the first message passed to the
+> component should be an empty string. Changes to the content of the component
+> will trigger subsequent announcements.
+
+### Impact
+
+This decision will impact existing usage of `aria-live`, `role="status"`, and
+`role="alert"` for the components listed below. These components will need to be
+updated to use the new approach, using one of the following approaches:
+
+- Use `announce()` or `announceFromElement()`
+- Use `AriaStatus` or `AriaAlert`
+
+In addition, we should make sure that `<live-region>` is successfully included
+in GitHub.
+
+#### Instances of `aria-live="polite"`
+
+- InputValidation
+- SelectPanel
+- TreeView
+
+#### Instances of `aria-live="assertive"`
+
+- InlineAutocomplete
+
+### Instances of `role="status"`
+
+- LiveRegion
+- TreeView
+- ActionList examples
+- Spinner
+
+### Instances of `role="alert"`
+
+None
+
+### Instances of `LiveRegion`
+
+- DataTable
+
+## Examples
+
+### Announce when content is shown
+
+```tsx
+import React from 'react'
+import {AriaStatus} from '@primer/react'
+
+function ExampleComponent() {
+  const [loading, setLoading] = React.useState(true)
+  if (loading) {
+    return <AriaStatus>Example loading message</AriaStatus>
+  }
+  return <Page />
+}
+```
+
+### Announce on content change
+
+```tsx
+import {AriaStatus} from '@primer/react'
+import {useState} from 'react'
+
+function ExampleComponent() {
+  const [count, setCount] = useState(0)
+  return (
+    <button
+      type="button"
+      onClick={() => {
+        setCount(count + 1)
+      }}
+    >
+      <AriaStatus>Count {count}</AriaStatus>
```

Comment:
If a page renders with `<ExampleComponent />`, would `Count 0` be announced, or would we only hear announcements starting from the content change?

Reply:
Based on your note above, it sounds like `Count 0` would be announced.

I'm curious what markup you would recommend if we want to visually show `Count 0` on render and don't want it announced, but still want to announce/visually convey subsequent changes 🤔 

Following the guidance in:
> In cases where they should always be present, then the first message passed to the
> component should be an empty string. Changes to the content of the component
> will trigger subsequent announcements.

it sounds like we'd need to do something like:

```.tsx
  Count {count}
  <VisuallyHidden><AriaStatus></AriaStatus></VisuallyHidden>
```
and manage both separately. This feels kind of messy to me, as opposed to just using `role="status"` which would work for this particular scenario 🤔

Reply:
> Based on your note above, it sounds like Count 0 would be announced.

Yes, you're spot on 👍 

The core challenge to me is that we can't distinguish between page load and a component rendering when deciding to make an announcement. Sometimes we want this component to announce when it's rendered and sometimes not (in particular during page load).

I think there are two ways to approach this:

1. Opt-in to announce on first render
2. Opt-out of announcing first render

So the `AriaStatus` would pick one as the default behavior and provide a prop to opt-out of that default. Based on what you shared, it seems like it'd be worth treating Option 1 as the default but I'm curious what you would say 👀

Reply:
Oh this is so tricky 😅. It could be unexpected for these React components to be called `AriaStatus` and `AriaAlert` but behave slightly different from how these native roles behave, but then again.. most people are not familiar with the nuances around how the native role behaves anyways.. I feel like either is fine as long as we make the behavior crystal clear to the consumer. Would it be overkill to make it a required prop? 🤔  

I would also be interested in @patrickhlauke thoughts, especially as he's also reviewing Primer components/docs and might have insights around what we designate as default behavior, and how it gets communicated to Primer consumers.

Another question - when a component is rendered with a `hidden` attribute, and then shown by removing the `hidden` attribute, is that considered "first render"? Or does this pattern usually not occur in React since things are typically conditionally rendered?

Reply:
> It could be unexpected for these React components to be called AriaStatus and AriaAlert

Totally agreed, thanks for bringing that up. Also wanted to make sure it's clear that if these abstractions aren't particularly helpful (especially compared to directly using `role="status"`) always happy to pursue a different route 👍 

Was revisiting what @patrickhlauke posted over at: https://github.com/w3c/aria/issues/2154 for this since it looks like we're running into this behavior where announcements are made only on _update_ and we are running into situations where we want to announce on insertion (such as a spinner).

With this in mind, we could have the default be what the corresponding role has:

- `status` will announce on update, by default
- `alert` will announce on insert, by default

And for `status` specifically we could have a prop like `announceOnShow` or similar that would provide that first announcement. If we prefer to just keep the roles mapped to `role="alert"` and `role="status"` we could also consider having a dedicated component just for this case, such as `Announce`, that will just announce its contents and that way we avoid having to make these kinds of considerations.

---

For the `hidden` question, great point. None of the components proposed are taking that into account and we could 100% add that in based on what you'd recommend. For example, each component can accept `hidden` and we only announce when `hidden={false}`.

Reply:
Made these updates over at: https://github.com/primer/react/pull/4673 FYI @khiga8